### ISSUE TITLE
qemu gets OOMkilled

### ISSUE THREAD

**Issue by dren-dk** [kubevirt/kubevirt#4560]
/kind bug

**What happened**:

Regardless of the size of the memory allocated to a VM it gets OOM killed.

I can work around the problem by manually specifying a larger resource limit than the VM memory size with:

* domain.memory.guest: 8Gi
* domain.resources.limits.memory: 9Gi
* domain.overcommitGuestOverhead: true

According to the routine that calculates the resource limit from the guest memory, the calculation is not entirely bullet proof, but
there doesn't seem to be any way to tweak the overhead calculation, such a mechanism would be very nice to have.

**What you expected to happen**:

I would expect the memory limit to be set high enough to allow qemu to run without triggering the OOMkiller.

**How to reproduce it (as minimally and precisely as possible)**:

I saw the problem with a windows 10 2004 installation with 3 virtio disks and a secret mounted as a sata cdrom.

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):

The server is running 0.35.0

Client Version: version.Info{GitVersion:"v0.29.0", GitCommit:"cbb7eaddb34d38da15868b19d84647f668a6134b", GitTreeState:"clean", BuildDate:"2020-05-05T03:33:18Z", GoVersion:"go1.12.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"{gitVersion}", GitCommit:"{gitCommit}", GitTreeState:"{gitTreeState}", BuildDate:"{buildDate}", GoVersion:"go1.13.14", Compiler:"gc", Platform:"linux/amd64"}

- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:17:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:13:49Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}

- VM or VMI specifications: 

This is the VM specification that doesn't get OOMkilled:

apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: win10-2004
spec:
  running: true
  template:
    metadata:
      annotations:
        debugLogs: "true"
      labels:
        vmName: win10-2004
        debugLogs: "true"
        
    spec:
      domain:
        memory:
          guest: 8Gi
        cpu:
          cores: 2
        resources:
          overcommitGuestOverhead: true
          requests:
            memory: 9Gi
          limits:
            memory: 9Gi
        devices:
          disks:
          - disk:
              bus: virtio
            name: boot
          - disk:
              bus: virtio
            name: service
          - disk:
              bus: virtio
            name: tools
          - cdrom:
              bus: sata
            name: secret-volume
        
          interfaces:
          - bridge: {}
            name: default
        machine:
          type: q35
      networks:
      - name: default
        pod: {}
      nodeSelector:
        overlay: overlay2
      volumes:
      - persistentVolumeClaim:
          claimName: win10-2004
        name: boot
      - containerDisk:
          image: ...
        name: service
      - containerDisk:
          image: ...
        name: tools
      - name: secret-volume
        secret:
          secretName: service-mode

- Cloud provider or hardware configuration: Bare metal.

- OS (e.g. from /etc/os-release): ubuntu 18.04

- Kernel (e.g. `uname -a`):  Linux haddock 4.15.0-122-generic #124-Ubuntu SMP Thu Oct 15 13:03:05 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**Comment by vladikr** [comment]
Thank you for filing this issue.

QEMU requires additional memory to function properly. It depends on the requested resources and various threads.

I see that you are using `overcommitGuestOverhead` API field.
By using this filed you are effectively requesting *not* to add any overhead to the memory requested for the container.
According to your configuration, you are also using a Guaranteed QOS, which means that the container in which QEMU runs is restricted in memory.

```
memory:
      guest: 9Gi
    cpu:
      cores: 2
    resources:
      overcommitGuestOverhead: true
      requests:
        memory: 9Gi
      limits:
        memory: 9Gi
```
In the above configuration, QEMU will allocate 9Gi to the guest. However, the moment QEMU will require more memory for its own operations, it will be killed.

Buy setting less memory for the guest and more memory for the container it's running in, you are effectively adding a 1Gi of overhead, as in the example you've posted:

```
memory:
      guest: 9Gi
    cpu:
      cores: 2
    resources:
      overcommitGuestOverhead: true
      requests:
        memory: 9Gi
      limits:
        memory: 9Gi
```
Such configuration will obviously work.

My suggestion is simply to stop adding `overcommitGuestOverhead: true` to VMIs with Guaranteed QOS.

In general, `overcommitGuestOverhead` was designed to be used with a Non-Guaranteed QOS VMIs (the vmis won't have a Resources.Limits.Memory specified) These VMI are not restricted in memory and can use more memory from the nodes as along and there are enough resources.

**Comment by rmohr** [comment]
I think there is one very save assumption in combination with `overcommitGuestOverhead: true`: If swap is disabled (which it normally should), this feature can't be used. In some corner cases it can make sense to run k8s with swap, but I would not recommend enabling it in general (the most obvious issue is that the kubelet can't detect memory pressure anymore).

**Comment by dren-dk** [comment]
I was not normally using overcommitGuestOverhead: true or memory.guest, I simply set 
 requests.memory: 9Gi
and then hoped that the overhead calculation would ensure that the VM would get a reasonable amount of RAM.

Was that wrong?

Should I set memory.guest and leave requests.memory to its own devices?

**Comment by dren-dk** [comment]
I have just removed overcommitGuestOverhead: true and messed about with options some more.

This configuration allocates exactly 8500 MB to qemu and OOMs, I was under the impression that the qemu process should get the amount of ram specified in the request, not the limit:
`
        resources:
          limits:
            memory: 8500Mi
          requests:
            memory: 8000Mi
`

This works to force qemu to use only 8000 Mi and it doesn't OOM on me:
`
        memory:
          guest: 8000Mi
        resources:
          limits:
            memory: 8500Mi
          requests:
            memory: 8000Mi
`

Leaving out the resources.limits.memory part of the VM, results in this VMI, because the default limit is 64 Gi in the namespace:
`
    memory:
      guest: 8000Mi
    resources:
      limits:
        cpu: '2'
        memory: 64Gi
      requests:
        cpu: '1'
        memory: 8000Mi
`

I want to ensure that there are never any pods without resource limits, so I have this LimitRange set for the namespace:
`
apiVersion: v1
kind: LimitRange
metadata:
  name: limit-range
spec:
  limits:
  - default:
      memory: 64Gi
      cpu: 2000m
    defaultRequest:
      memory: 2Gi
      cpu: 1000m
    max:
      memory: 70Gi
      cpu: 4000m
    type: Container
`

Leaving out memory.guest too, results in a qemu memory parameter of -m 65536 and it ends up OOM'ing.

The VMI resource limits when it OOMs is:
`
   resources:
      limits:
        cpu: '2'
        memory: 64Gi
      requests:
        cpu: '1'
        memory: 8000Mi
`


I ended up with this in the VM specification:
`
spec.domain.memory.guest: 8Gi
spec.domain.resources.requests.memory: 8Gi
`
The default memory limit from the LimitRange, then means that the memory limit of the VMI ends up at 64 Gi.

It's somewhat unsatisfying to not have a tighter memory limit on the pod by default, but it's workable for now and I guess a VM workload will have much less risk of exploding over the limit imposed by qemu.


Would it not be reasonable to use these defaults, unless specified in the VM manifest?:

* Set spec.domain.memory.guest.memory to spec.domain.resources.requests.memory.
* Set spec.domain.resources.requests.memory to spec.domain.memory.guest.memory.
* Set spec.domain.resources.limits.memory to spec.domain.resources.requests.memory + calculated overhead.

**Comment by rmohr** [comment]
> * Set spec.domain.memory.guest.memory to spec.domain.resources.requests.memory.

If no limit is set, it will use it.

> * Set spec.domain.resources.requests.memory to spec.domain.memory.guest.memory.

Yes, here we right now take the limit over requests. I see that this can be impractical when limits and requests are specified on the namespace, but no guest memory is explicitly set. It may be indeed be better to change that.

> * Set spec.domain.resources.limits.memory to spec.domain.resources.requests.memory + calculated overhead.

The overhead is always added to what people specify on the VMI, since also being constantly over your requests can sooner or later lead to evictions. The background is that we don't think that it is practical for people to have to experiment on every VM to find the right values (note that you can). If you want to do custom calibration, you can indeed use `overcommitGuestOverhead` without swap and experiment for yourself.


To sum up it may make sense to take requests over limits on the VMI due to cases like namespace range-limits.

**Comment by rmohr** [comment]
> To sum up it may make sense to take requests over limits on the VMI due to cases like namespace range-limits.

Actually, we don't enable ballooning yet, but when ballooning is present one would actually expect the opposite, because one could never go over what is requested and adding a limit higher than a request does not make much sense then. Maybe we should only stay on the save side when the values are coming from the range-limit of the namespace and still choose the limit if it was explicitly specified.

**Comment by mlsorensen** [comment]
There might be some work to do on calculating the guest memory, as I have been seeing VMs get OOMKilled as well without using overcommit. At first I thought perhaps that the node was experiencing memory pressure due to other pods having a very high limit but the node was carved up based on requests. If pods started going far beyond their request, that could cause the node to start looking for processes to kill. However after setting the requests and limits on the VM to match and confirming "Guaranteed" QoS on the virt-launcher pod, so it would theoretically be one of the last processes it looked at to remove, we still see OOM in some cases that look like exceeding pod resources.

In this last case I looked at, we just touched the pod limit:

```
memory: usage 196187536kB, limit 196187536kB, failcnt 55
```

For now we are trying the under commit feature by adding a spec.domain.memory.guest field that is lower than the matching requests and limits.
